src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlIds.java

```diff
@@ -491,4 +544,81 @@ public HtmlId forHeading(CharSequence headingText, Set<String> headingIds) {
         }
         return HtmlId.of(idValue);
     }
+
+    private final Map<ExecutableElement, HtmlId> ids = new HashMap<>();
+
+    /*
+     * Returns an id to a constructor or a method, from a centralised
+     * registry. Use to get an anchor to a constructor or method.
+     *
+     * The goal is to provide coordination, not cache.
+     */
+    private HtmlId forExecutable(ExecutableElement e) {
+        HtmlId htmlId = ids.get(e);
+        if (htmlId != null)
+            return htmlId;
+        if (e.getKind() != ElementKind.CONSTRUCTOR
+                && e.getKind() != ElementKind.METHOD)
+            throw new IllegalArgumentException(String.valueOf(e.getKind()));
+        var vmt = configuration.getVisibleMemberTable((TypeElement) e.getEnclosingElement());
+        var ctors = vmt.getVisibleMembers(VisibleMemberTable.Kind.CONSTRUCTORS);
+        var methods = vmt.getVisibleMembers(VisibleMemberTable.Kind.METHODS);
+        // for whatever reason annotation methods are not of Kind.METHODS
+        var otherMethods = vmt.getVisibleMembers(VisibleMemberTable.Kind.ANNOTATION_TYPE_MEMBER);
```

Comment:
I'm surprised you need to worry about annotation type members here -- annotation types cannot have type arguments, and so the "simple" id should always be sufficient.

Reply:
True, type parameters are not an issue for annotation interface methods, which [are not allowed to have any parameters][], type or otherwise. However, the code that prints annotations for method signatures does not know that and uses `forMember`, which is applicable to any executable member, of annotation or otherwise.

<img width="356" alt="a screenshot from the generated API Documentation for testNewAndDeprecated" src="https://github.com/openjdk/jdk/assets/32523691/fc808159-7f3f-4a2a-bb25-41474c3b5833">

In principle, I could remove that `vmt.getVisibleMembers(VisibleMemberTable.Kind.ANNOTATION_TYPE_MEMBER)` and the annotation member will be caught by the ["safety net"](https://github.com/openjdk/jdk/pull/18519/files#diff-22d9182196ae739a6de9c29801bb3ca788992b0cbf44564b2aeda2018a7b78e1R611-R621).

Since we are here, there's a `forMember` overload used by `AnnotationTypeMemberWriter`, I probably should remove it for consistency. Thoughts?

[are not allowed to have any parameters]: https://docs.oracle.com/javase/specs/jls/se22/html/jls-9.html#jls-9.6.1

Reply:
If you use `forMember` on an `ExecutableElement` whose enclosing element is an annotation type interface, you know there cannot be any type parameters.

Reply:
Right, but some accommodation/special-casing for annotations will be there anyway:

  * "safety net",
  * `getVisibleMember(VisibleMemberTable.Kind.ANNOTATION_TYPE_MEMBER)`
  * `e.getEnclosingElement().getKind() == ElementKind.ANNOTATION_TYPE`

I have no preference, let alone strong opinion on either of those options. If you are concerned with "performance", which I know you might, we could choose the latter option.

Reply:
I was not specifically concerned about performance in this case. It was more a case of code clarity/confusion (possibly my confusion).

So I guess the choice is one of:
* a common code path for all executable elements, whatever their container, regardless of whether there may be any risk of a clash of member ids caused by type parameters or anything else
* fast-track some code paths when you know there are no type parameters involved